### DISCUSSION TITLE
Can i download a .swf file through the browser path bar when it's not a legal SWF file?

### DISCUSSION THREAD

**Discussion by shaucky** [ruffle-rs/ruffle#18161] [General] [answered]
There're some SWF which bytes have been compressed, they are a part of a web page game.
 
Now i can't paste the path to download it because Ruffle would try to play it. And its bytes have been changed, so it seems not a legal SWF file to Ruffle. At the same time, Ruffle do not give a choice to download the file.

**Comment by Lord-McSweeney** [comment]
You can disable Ruffle's playing of SWF files in the browser by going to the extension settings and disabling "Play SWF files in browser instead of downloading". Alternatively, you can simply turn off the Ruffle extension.

**Reply by shaucky** [reply]
Okay, but i didn't see a setting like "Play SWF files in browser instead of downloading", so i turn off the extension. I'm using MS Edge. Thanks.
 
![FD4A91725669BE9D8AFDB132BBC23898](https://github.com/user-attachments/assets/213ecbb2-afa9-44c8-9610-47f7ef6c538c)

**Reply by danielhjacobs** [reply]
Click "Extension Options" (扩展选项), I believe in your screenshot it's the second-to-last link.

On the page to which clicking that takes you, disable "Play SWF files in browser instead of downloading" (在浏览器中直接播放 SWF 文件，而不是下载):

![image](https://github.com/user-attachments/assets/e2df87b0-54cc-4aad-a40f-a9cb7b3df0b9)

**Reply by shaucky** [reply]
Wow, thanks for your patiently translating into my language! But sadly there're only 3 options on my screen…
 
![BD72798FA0636D8CC8E368E45DC76F6B](https://github.com/user-attachments/assets/03c1bbf8-3e58-4616-bb84-1dd0de3737ff)

**Reply by shaucky** [reply]
Well, at the moment of refreshing this settings page, i saw the English of the fourth option flashing. I guess the translation for the corresponding option was not configured, so this option was hidden after translation. But i found through developer tools that this option exists and is translated, but comes with a hidden class. Interesting.
 
![5C99AB744DA2FD1E69B904A6C19A5FE0](https://github.com/user-attachments/assets/2e261afe-a8ed-4f54-b6f1-78a849d21492)

**Reply by danielhjacobs** [reply]
Hi, that means you're using a version of Microsoft Edge prior to version 128. We rely on https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/RuleCondition#responseheaders and are only able to "play SWF files in browser instead of downloading" on browsers that have that implemented, which are currently just browsers based on Chromium 128+ (Microsoft Edge uses Chromium as its browser engine). What is strange is that on those older browsers, your issue of "i can't paste the path to download it because Ruffle would try to play it" shouldn't occur, since a direct SWF link doesn't play using Ruffle on those older browsers.

You can enable "在右键菜单中显示SWF下载选项", and then when you right-click a file that loaded with Ruffle you can download it with the option "下载 .swf", but that will only work if Ruffle succeeds in loading the file.

With that in mind, the only possibilities I can think of for why this issue is occurring for you is either the SWF link is really an HTML page that loads with Ruffle but is meant to look like an SWF link (the Internet Archive does that) or, if you're in China, Microsoft Edge's own engineering team has integrated Ruffle into the browser itself (I believe they are working on that for the Chinese market but I don't know how far that has gone).

**Reply by shaucky** [reply]
> Hi, that means you're using a version of Microsoft Edge prior to version 128.

 
My MS Edge is usually auto update so it's version almost keeps the new version.
 
![CEBA18FEE9D3299F97B79396A2C07343](https://github.com/user-attachments/assets/997404c8-c2a0-4a34-877a-8330501c51be)
 
It's sure that using Chromium. If i input a normal SWF file path, Ruffle will load and play the file. Like the screenshot:
 
![93DFE786E12A1B57E5DDD1F01192EAA0](https://github.com/user-attachments/assets/7511ad85-21dd-452d-a8ee-30dd294a3723)
 
And then, if i input a SWF file path which file has been changed bytes. What behavior is like this:
 
![2A1E69B321DA564FBBB1F8F993AB30D8](https://github.com/user-attachments/assets/88d6428a-75eb-4778-91c4-0bb7053bdd77)
 
> the only possibilities I can think of for why this issue is occurring for you is either the SWF link is really an HTML page that loads with Ruffle but is meant to look like an SWF link
 
The situation here is that this SWF is part of the core of this ancient Flash web page game, have been processed by the game official to may because increase the difficulty for players to find undisclosed data. If i remember correctly, the bytes of the original normal SWF file were compressed by deflate and then moved backwards by 8 bytes.
 
> I believe they are working on that for the Chinese market but I don't know how far that has gone
 
On this side, i don't know the integration of Ruffle and MS Edge (This may because I more focus on Adobe AIR). I think i actively got Ruffle from Microsoft-Edge-Extensions. However, in China, Adobe Flash Player has an independent agent who provides region specific Adobe Flash Player version 34.
 
Thanks.

**Reply by danielhjacobs** [reply, marked as answer]
Try uninstalling and reinstalling the official Ruffle Edge extension from https://microsoftedge.microsoft.com/addons/detail/ruffle-flash-emulator/pipjjbgofgieknlpefmcckdmgaaegban. It's possible a variable called "responseHeadersUnsupported" was set to false in https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/sync by Ruffle before you ever had Edge version 128 installed and the variable hasn't been updated since. After re-installing, the option "在浏览器中直接播放 SWF 文件，而不是下载" may appear.

**Reply by shaucky** [reply]
Yeah, the reinstalled Ruffle shows the 4th option in it's options page. I guess you're right in a way.
 
Although i don't think i had an old version of MS Edge before i install Ruffle, because my computer came with Win 11, and Edge had already provided Chromium based version before the release of Win 11. Anyway, the problem has indeed been solved.
 
Thanks for your kinds.

**Comment by n0samu** [comment]
If you can load the game in the Ruffle desktop app, you can use its "Show known movies" functionality in its Debug Tools to save the decrypted SWF. Otherwise, you can save a memory dump of the browser's memory and extract the decrypted SWF from the memory dump.

**Reply by shaucky** [reply]
I haven't tried the desktop version of Ruffle yet, but this Flash web page game has a login SDK that interacts with the carrier HTML page, so i think this solution may not be feasible, at least not in the desktop version of Adobe Flash Player.
 
Waiting for the decrypted SWF file to be loaded into memory may be one way to obtain it, but i lack experience.
 
Thanks.
